File: RestoreIP.php
<?php

class Solution
{

    /**
     * @param String $s
     * @return String[]
     */
    function restoreIpAddresses($s)
    {
        if (strlen($s) > 12) {
            return [];
        }
        $res = [];

        $this->dfs(0, 0, '', $s,$res);

        return $res;
    }

    function dfs($i, $dots, $curIP, $s, &$res)
    {
        if ($dots == 4 && $i  == strlen($s)) {
            $res[] = trim($curIP, '.');
            return;
        }

        if ($dots > 4) {
            return;
        }


        for ($j = $i; $j < strlen($s); $j++) {
            $len = $j -$i +1;
            $curstr = substr($s, $i, $len);

            if ((int)$curstr < 256 && ($j == $i || $curstr[0] != 0)) {
                $this->dfs($j + 1, $dots + 1, $curIP . $curstr . ".", $s,$res);
            }
        }
    }
}

$sln = new Solution();

var_export($sln->restoreIpAddresses("101023"));

File: RestoreIP.py
class Solution:
    def restoreIpAddresses(self, s: str):

        res =[]

        if len(s) > 12: 
            return res
        
        def backtack(i,dots,curIP):
            if dots == 4 and i == len(s):
                res.append(curIP[:-1])
                return
            if dots > 4:
                return
            l = len(s)
            for j in range(i, min(i +3, l)):
                curstr = int(s[i:j+1])
                if curstr < 256 and (i == j or s[i] != '0'):
                    backtack(j+1, dots + 1,curIP + s[i:j+1]+".")
        backtack(0,0,"")
        return res

sln = Solution()
print(sln.restoreIpAddresses('101023'))